### DISCUSSION TITLE
How to set up 404 handler in case of a custom express app

### DISCUSSION THREAD

**Discussion by Isaac-Leonard** [RobinTail/express-zod-api#155] [Q&A] [answered]
Would it be possible to expose the default 404 handler so it can be used when using a custom express instance to provide a consistent api interface with minimal effort?

**Comment by RobinTail** [comment]
Hello @Isaac-Leonard ,

the handler for 404 (Not found) errors can be globally configured when using `createServer()` method (not for custom express instance):
```typescript
import {createConfig, createResultHandler} from 'express-zod-api';

const myResultHandler = createResultHandler({
  ...,
  handler: ({error, input, output, request, response, logger}) => {
    // your implementation, for example:
    if (error && error.status === 404) {
      response.status(error.status).end(error.message);
    }
  }
});

const config = createConfig({
  errorHandler: myResultHandler,
  ...
});

const routing = {...};
createServer(config, routing);
```

In case of the request to a path that is not declared in `routing` it calls the `myResultHandler` with an `error` argument which is `HttpError` with a status code `404`, so you can handle this case within.

There is a method `attachRouting` designed specifically for a custom express instance/app that only connects Endpoints to their paths and methods. So it gives you a power on the app customization, but, as it should be known from the Spiderman lore, the power comes with the responsibility. In this regard you can add your 404 handler right next to the `attachRouting()` call:

```typescript
import * as express from 'express';
import {createConfig, attachRouting} from 'express-zod-api';

const app = express();
const config = createConfig({app, ...});
const routing = {...};

attachRouting(config, routing);

app.use((req, res) => {
  // your 404 handler implementation, for example:
  res.status(404).end(`Path ${req.path} not found.`);
});

app.listen();
```

So the library provides two options:
- either it creates the express app itself and you can configure the error handler;
- or you create your own express app, connect your endpoints and implement your own function for error handling as a part of your customization. This topic is also covered in [Readme](https://github.com/RobinTail/express-zod-api#your-custom-server).

I hope one of these solutions will suit your needs. Please let me know.

If you have any suggestions on how this design can be improved, please also let me know.

**Comment by RobinTail** [comment]
Hello again @Isaac-Leonard 

> to expose the default 404 handler

The easiest solution for using the existing `ResultHandler` for handling 404 errors in case of a custom express app would be calling its `.handler()` method with a bunch of parameters including the error itself:

```typescript
// ...
attachRouting(config, routing);

app.use((request, response) => {
  // any ResultHandler.handler() of your choice
  defaultResultHandler.handler({
    request, response, 
    error: createHttpError(404, `${request.path} not found`),
    logger: createLogger(config.logger), // or config.logger in case of custom logger
    input: null,
    output: null
  });
});

app.listen();
```

So basically it's exposed. The problem currently is the logger instance. If the custom logger is not used then it only exists within `attachRouting()` and not exposed.

**Comment by RobinTail** [comment]
I think that I can make `attachRouting` to return something. Currently it's void, but it can be anything.
Thank you for the idea @Isaac-Leonard. Still need to think about it.

**Comment by RobinTail** [comment]
@Isaac-Leonard , I made a PR #154 which I believe will resolve the issue and provide you with at least two options for your case.

**Comment by RobinTail** [comment]
The feature will be released in version 2.6.0.

**Comment by RobinTail** [comment, marked as answer]
@Isaac-Leonard ,

the feature has been released in version 2.6.0. Please upgrade and so can do the following:

```typescript
const {notFoundHandler} = attachRouting(config, routing);

app.use(notFoundHandler);

app.listen();
```

**Comment by Isaac-Leonard** [comment]
Hi @RobinTail
This solution is perfect thank you.
Sorry I didn't respond earlier, GitHub emails have been getting sent to junk for some reason.
This update is very helpful though.
Best regards Isaac
